lib/signalfx.js

```diff
@@ -0,0 +1,334 @@
+    useProtocolBuffers = false;
+  }
+
+// ============================================================================================
+//                            BaseSignalFx Signal Fx class
+// ============================================================================================
+  function BaseSignalFx(apiToken, ingestEndpoint,
+                        apiEndpoint, timeout,
+                        batchSize, userAgents) {
+    this.apiToken = apiToken;
+    this.ingestEndpoint = ingestEndpoint || DEFAULT_INGEST_ENDPOINT;
+    this.apiEndpoint = apiEndpoint || DEFAULT_API_ENDPOINT;
+    this.timeout = timeout || DEFAULT_TIMEOUT;
+    this.batchSize = Math.max(1, (batchSize ? batchSize : DEFAULT_BATCH_SIZE));
+    this.userAgents = userAgents || null;
+  }
+
+  BaseSignalFx.prototype.send = function (cumulativeCounters, gauges, counters) {
+    if (!cumulativeCounters && !gauges && !counters) {
+      return;
+    }
+
+    return {
+      cumulative_counter: cumulativeCounters,
+      gauge: gauges,
+      counter: counters
+    };
+  };
+
+  BaseSignalFx.prototype.sendEvent = function (eventType, dimensions, properties) {
+    return {
+      eventType: eventType,
+      dimensions: dimensions || [],
+      properties: properties || []
+    };
+  };
+
+
+// ============================================================================================
+//                          Client Signal Fx class - base API client
+// ============================================================================================
+  /**
+   * SignalFx API client.
+   * This class presents a programmatic interface to SignalFx's metadata and
+   * ingest APIs. At the time being, only ingest is supported; more will come
+   * later.
+   * @param apiToken
+   * @constructor
+   */
+  function SignalFxClient(apiToken) {
+    BaseSignalFx.apply(this, arguments);
+
+    this.HEADER_API_TOKEN_KEY = 'X-SF-Token';
+    this.HEADER_USER_AGENT_KEY = 'User-Agent';
+    this.HEADER_CONTENT_TYPE = 'Content-Type';
+
+    this.INGEST_ENDPOINT_SUFFIX = 'v2/datapoint';
+    this.API_ENDPOINT_SUFFIX = 'v1/event';
+
+    this.queue = [];
+    this.asyncRunning = false;
+  }
+
+  SignalFxClient.prototype = Object.create(BaseSignalFx.prototype);
+  SignalFxClient.prototype.constructor = BaseSignalFx;
+
+  /**
+   * Send the given metrics to SignalFx.
+   *
+   * @param cumulativeCounters (list): a list of dictionaries representing the cumulative counters to report.
+   * @param gauges (list): a list of dictionaries representing the gauges to report.
+   * @param counters (list): a list of dictionaries representing the counters to report.
+   */
+  SignalFxClient.prototype.send = function (cumulativeCounters, gauges, counters) {
+    var data = BaseSignalFx.prototype.send.apply(this, arguments);
+
+    if (data === null || data === undefined) {
+      return;
+    }
+
+    var _this = this;
+    for (var metricType in data) {
```

Comment:
You're transforming your statically defined args (cumulativeCounters, guages, counters) to a map ('data'), and then iterate over the properties of the map to perform operations, and this seems like unnecessary work.

This function could be reduced to:

    function processDataPoints(metricType, dataPoints) {
      if(!dataPoints) return;
      if(!Array.isArray(dataPoints)) {
        throw new TypeError // ...
      }
    
      dataPoints.forEach(function(dp) { 
        _this.addToQueue(metricType, dp); 
      });
    }
    
    processDataPoints('cumulative_counter', cumulativeCounters);
    processDataPoints('guage', guages);
    processDataPoints('counter', counters);

This way there is no awkward transformation of the arguments in to another datastructure just to iterate over it. Better yet, you could have send be an internal function which basically does what processDataPoints does along with the startAsyncSend and create higher level sugar functions for .sendCumulativeCounters .sendGuages .sendCounters so that client of the library don't have to remember the arbitrary parameter positions.